### PR TITLE
Add code to generate training datasets

### PR DESCRIPTION
**Pull request by Naqu6**

### CODE REVIEW

File: videoretrieval/train/language_model.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    def inference(ids):
+        return language_model(ids)
+
+    def wrapper(video_id, ids, tokens):
+        contextual_embeddings = tf.py_function(inference, [ids], tf.float32)
+        return video_id, contextual_embeddings, tokens
+
+    return wrapper
+
+def generate_contextal_embeddings(language_model, dataset):
```
Comment (sbucak) [inline videoretrieval/train/language_model.py:69]: contextal -> contextual

File: videoretrieval/tests/encoder_datasets_tests.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                expected_feature, expected_missing = \
+                    self.mock_precomputed_features[feature_index][video_id]
+                
+                self.assertTrue(np.array_equal(
+                    feature.numpy(), expected_feature.numpy()))
+                self.assertEqual(missing, expected_missing)
+
+
+    def test_update_dataset_shape_wrapper(self):
+        pass
```
Comment (sbucak) [inline videoretrieval/tests/encoder_datasets_tests.py:125]: smooth test :)

File: videoretrieval/cache/language_model_cache.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+def get_feature(value):
+    """Gets a tf.train.Feature from the parameter value."""
+    bytes_list = tf.train.BytesList(value=[value.numpy()])
+    return tf.train.Feature(bytes_list=bytes_list)
+
+def get_records_directory(dataset, language_model, split):
+    """Gets a path to cache the data from the dataset/model/split."""
+    dataset_name = dataset.dataset_name
+    language_model_name = language_model.name
+
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/cache/language_model_cache.py:43]: Got it

File: videoretrieval/cache/language_model_cache.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        tokens: the number of tokens created from the original caption.
+
+    Returns:
+        A protobuf serialized as a string.
+    """
+    video_id_feature = get_feature(video_id)
+
+    # Accessing contextual_embeddings[0] to gets rid of the extra dimension.
+    serialized_embedding = tf.io.serialize_tensor(
+        contextual_embeddings[0, :tokens])
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/cache/language_model_cache.py:65]: I've added an assertion on the line above.

File: videoretrieval/cache/language_model_cache.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        video_id = example["video_id"]
+
+        contextual_embeddings = tf.io.parse_tensor(
+            example["serialized_embeddings"], tf.float32)
+
+        embedding_length = tf.numpy_function(
+            get_embedding_length, [contextual_embeddings], tf.int64)
+
+        if embedding_length >= text_max_length:
+            return (video_id, contextual_embeddings[:text_max_length])
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/cache/language_model_cache.py:192]: Correct.

File: videoretrieval/train/encoder_datasets.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+import cache
+import tensorflow as tf
+import numpy as np
+
+def replace_video_id_with_expert_features_wrapper(precomputed_features):
+    """Returns a function that adds precomputed features to an example.
+
+    Arguments:
+        precomputed_features: an array of dicts, where each dict maps from a
+            video id to a precomputed feature.
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/train/encoder_datasets.py:28]: My documentation was unclear. I've updated it :)

File: videoretrieval/train/encoder_datasets.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+
+    def get_expert_features(video_id_encoded):
+        """Gets the features for a given video id"""
+        expert_features = []
+        missing_modalities = []
+
+        video_id = video_id_encoded.decode("utf-8")
+
+        for feature_dict in precomputed_features:
+            features, data_exists = feature_dict[video_id]
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/train/encoder_datasets.py:46]: It should always exist: we should precomputed features for each feature in precomputed features. I've added an assertion.

File: videoretrieval/train/encoder_datasets.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+
+    return wrapper
+
+def update_dataset_shape_wrapper(experts, language_model):
+    """Updates the shapes of expert features and text embedding a given dataset.
+
+    Arguments:
+        experts: a list of experts (type BaseExpert) used in the dataset.
+        language_model:
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/train/encoder_datasets.py:70]: Whoops! All fixed.

File: videoretrieval/train/encoder_datasets.py [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+def update_dataset_shape_wrapper(experts, language_model):
+    """Updates the shapes of expert features and text embedding a given dataset.
+
+    Arguments:
+        experts: a list of experts (type BaseExpert) used in the dataset.
+        language_model: 
+
+    Returns: a function that takes in video id, expert features, contextual
+        embeddings, and missing modalities as parameters and assigns a shape to
+        the contextual embeddings and the expert features, then returns the a
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/train/encoder_datasets.py:74]: Yup!

File: videoretrieval/train/language_model.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import tensorflow as tf
+from cache import cache_language_model_embeddings 
+
+
+def get_encode_function(language_model):
+    """Wraps a function that uses a language model's tokenizer to encoder text.
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/train/language_model.py:22]: 👍

File: videoretrieval/train/language_model.py (hunk 1 above)
Comment (Naqu6) [reply videoretrieval/train/language_model.py:69]: 👍

File: videoretrieval/tests/encoder_datasets_tests.py (hunk 2 above)
Comment (Naqu6) [reply videoretrieval/tests/encoder_datasets_tests.py:125]: Thanks haha

**Comment by Naqu6** [comment]
Of course!

**Pull request merged**